Add Example.md file in ServiceNow Connector (#1291)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/servicenow-connector/servicenow-connector-example.md

# ServiceNow Connector Example

The ServiceNow connector allows you to access the ServiceNow REST API through WSO2 EI. ServiceNow is a software platform that supports IT service management and automates common business processes. This software as a service (SaaS) platform contains a number of modular applications that can vary by instance and user. 

> It focuses on service-orientation toward the tasks, activities, and processes.

Reviewer: The introduction of what is `serviceNow` does not give a overall idea what it is and what it does. 

1. This video has what it can do. https://www.youtube.com/watch?v=g1g9iJc6WyY
2. We also need to say what this connector does. Not only what serviceNow does. The connector can be used to trigger workflows? Or can it create workflows? 
3. Add a link to servicenow site  and the API we support. 
https://developer.servicenow.com/dev.do#!/reference/api/orlando/rest/c_TableAPI

Author: Done

## What you'll build

> This example explains how to use ServiceNow Connector to create records in a table and retrieve its information. Assume your organization uses ServiceNow support and you need to create an incident. In order to do that we can use the tableAPI which is a Rest API. This can be easily done using WSO2 ServiceNow connector. We can use the following API which is designed using the ServiceNow connector to create record in the incident table as well as to read the created incident ticket. 

Reviewer: a table >> incident table of `serviceNow`?

Reviewer: Whenever integration needs to raise an incident at `serviceNow`, above EI API can be called with required information.

Author: Done

It will have two HTTP API resources, which are `postRecord` and `readRecord`. 

* `/postRecord`: It creates a new record in the existing incident table in ServiceNow instance 

* `/readRecord `: It reads the detailed information about the created incident record in the incident table.

>     <img src="/assets/img/connectors/serviceNow.png" title="Adding a Rest API" width="800" alt="Adding a Rest API"/>

Reviewer: Should image come under `/readRecord`. Pls verify with how MkDocs render it.

Author: Changed.

## Configure the connector in WSO2 Integration Studio

> Follow these steps to set up the ESB Solution Project and the Connector Exporter Project. 

Reviewer: Note that 'ESB Solution Project' now renamed as 'Integration Project' now.

Author: Done.

{!references/connectors/importing-connector-to-integration-studio.md!} 

1. First let's create postRecord sequence and ReadRecord sequences. Right click on the created ESB Solution Project and select, -> **New** -> **Sequence** to create the Sequence. 
    <img src="/assets/img/connectors/add-sequence.png" title="Adding a Sequence" width="800" alt="Adding a Sequence"/>

> 2. Provide the Sequence name as PostRecord. You can go to the source view of the xml configuration file of the API and copy the following configuration. 

Reviewer: xml >> XML

Author: Changed

          <sysparmInputDisplayValue>true</sysparmInputDisplayValue>
          <number>34</number>
          <shortDescription>{$ctx:shortDescription}</shortDescription>
          <active>true</active>
          <approval>owner</approval>
          <category>inquiry</category>
          <contactType>{$ctx:contactType}</contactType>
      </servicenow.postRecord>
      <property expression="json-eval($.result.sys_id)" name="sysId" scope="default" type="STRING"/>
      </sequence>
    ```
3. Create the ReadRecord sequence as below. 
  ```
    <?xml version="1.0" encoding="UTF-8"?>
    <sequence name="ReadRecord" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
        <servicenow.init>
            <serviceNowInstanceURL>https://dev55707.service-now.com</serviceNowInstanceURL>
            <username>admin</username>
            <password>Diazo123@</password>
        </servicenow.init>
        <servicenow.getRecordById>
            <sysId>{$ctx:sysId}</sysId>
            <tableName>incident</tableName>
        </servicenow.getRecordById>
    </sequence>
  ```
  4. Now right click on the created ESB Solution Project and select, -> **New** -> **Rest API** to create the REST API. 

>     <img src="/assets/img/connectors/adding-an-api.png" title="Adding a Rest API" width="800" alt="Adding a Rest API"/>

Reviewer: Do we need this image? @samgnaniah idea is having a lot of images in the doc make it confusing to the user.

Author: Removed

>   5. Provide the API name as ServiceNowAPI and the API context as `/servicenow`. You can go to the source view of the xml configuration file of the API and copy the following configuration. 

Reviewer: xml >> XML

Author: Changed

                  <property expression="json-eval($.contactType)" name="contactType" scope="default" type="STRING"/>
                  <sequence key="PostRecord"/>
                  <respond/>
              </inSequence>
              <outSequence/>
              <faultSequence/>
          </resource>
          <resource methods="POST" uri-template="/readRecord">
              <inSequence>
                <property expression="json-eval($.sysId)" name="sysId" scope="default" type="STRING"/>
                  <sequence key="ReadRecord"/>
                  <respond/>
              </inSequence>
              <outSequence/>
              <faultSequence/>
          </resource>
      </api>

    ```

{!references/connectors/exporting-artifacts.md!}

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}

## Testing

### Post Record Operation

1. Create a file called data.json with the following payload. 
    ```
    {
        "shortDescription":"Incident type: L2",
        "contacttype":"email"
    }

>     ```

Reviewer: Reader will wonder what should come as `contacttype `. Shall we point to the reference doc here?

Author: Added

# ServiceNow Connector Example

ServiceNow is an application platform as a service, which is a cloud-based computing model that provides the infrastructure needed to develop, run, and manage applications. It offers activities of an organization such as data collection, storage, workflow automation, and reporting through a single user interface. This software as a service (SaaS) platform contains a number of modular applications that can vary by instance and user. It focuses on service-orientation toward the tasks, activities, and processes.

> The WSO2 ServiceNow connector allows you to access the ServiceNow REST API through WSO2 EI. Using ServiceNow connector you can work with Aggregate API, Import Set API and Table API in ServiceNow. You can further read about ServiceNow REST APIs from [here.](https://developer.servicenow.com/dev.do#!/reference/api/orlando/rest/c_TableAPI)

Reviewer 2: ```suggestion
The WSO2 ServiceNow connector allows you to access the ServiceNow REST API through WSO2 EI. Using ServiceNow connector you can work with Aggregate API, Import Set API and Table API in ServiceNow. You can further read about ServiceNow REST APIs from [here](https://developer.servicenow.com/dev.do#!/reference/api/orlando/rest/c_TableAPI).
```

Author: Done

## What you'll build

> This example explains how to use ServiceNow Connector to create records in a table and retrieve its information. Assume your organization uses ServiceNow support and you need to create an incident. In order to do that we can use the tableAPI which is a Rest API. This can be easily done using WSO2 ServiceNow connector. Whenever integration needs to raise an incident at ServiceNow, above EI API can be called with required information.

Reviewer 2: ```suggestion
This example explains how to use ServiceNow Connector to create records in a table and retrieve its information. Assume your organization uses ServiceNow support and you need to create an incident. In order to do that we can use the tableAPI, which is a Rest API. This can be easily done using WSO2 ServiceNow connector. Whenever you need to raise an incident at ServiceNow, the above API can be called with the required information.
```

Author: Done

It will have two HTTP API resources, which are `postRecord` and `readRecord`. 

<img src="/assets/img/connectors/serviceNow.png" title="ServiceNow Scenario" width="800" alt="ServiceNow Scenario"/>

> * `/postRecord`: It creates a new record in the existing incident table in ServiceNow instance 

Reviewer 2: ```suggestion
* `/postRecord`: It creates a new record in the existing incident table in the ServiceNow instance.
```

Author: Done

* `/readRecord `: It reads the detailed information about the created incident record in the incident table.

## Setting up the environment 

> Please follow the steps mentioned at [Setting up Servicenow Instance](settingup-servicenow-instance.md) document in order to create a Servicenow Instance and obtain the credentials. 

Reviewer 2: ```suggestion
Please follow the steps mentioned in the [Setting up Servicenow Instance](settingup-servicenow-instance.md) document in order to create a ServiceNow Instance and obtain the credentials. 
```

Author: Done

Keep them saved to be used in the next steps.  

## Configure the connector in WSO2 Integration Studio

Follow these steps to set up the Integration Project and the Connector Exporter Project. 

{!references/connectors/importing-connector-to-integration-studio.md!} 

> 1. First let's create postRecord sequence and ReadRecord sequences. Right click on the created ESB Solution Project and select, -> **New** -> **Sequence** to create the Sequence. 

Reviewer 2: Need to update this. It is called Integration Project in the latest tooling pack.

Author: Done

          <sysparmView>short_description,number,sys_id</sysparmView>
          <sysparmInputDisplayValue>true</sysparmInputDisplayValue>
          <number>34</number>
          <shortDescription>{$ctx:shortDescription}</shortDescription>
          <active>true</active>
          <approval>owner</approval>
          <category>inquiry</category>
          <contactType>{$ctx:contactType}</contactType>
      </servicenow.postRecord>
      <property expression="json-eval($.result.sys_id)" name="sysId" scope="default" type="STRING"/>
      </sequence>
    ```
3. Create the ReadRecord sequence as below. 
  ```
    <?xml version="1.0" encoding="UTF-8"?>
    <sequence name="ReadRecord" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
        <servicenow.init>
            <serviceNowInstanceURL>https://dev55707.service-now.com</serviceNowInstanceURL>
            <username>admin</username>
            <password>Diazo123@</password>
        </servicenow.init>
        <servicenow.getRecordById>
            <sysId>{$ctx:sysId}</sysId>
            <tableName>incident</tableName>
        </servicenow.getRecordById>
    </sequence>
  ```

>   4. Now right click on the created ESB Solution Project and select, -> **New** -> **Rest API** to create the REST API. 

Reviewer 2: ```suggestion
  4. Now right click on the created Integration Project and select **New** -> **Rest API** to create the REST API. 
```

Author: Done

          <resource methods="POST" uri-template="/postRecord">
              <inSequence>
                  <property expression="json-eval($.shortDescription)" name="shortDescription" scope="default" type="STRING"/>
                  <property expression="json-eval($.contactType)" name="contactType" scope="default" type="STRING"/>
                  <sequence key="PostRecord"/>
                  <respond/>
              </inSequence>
              <outSequence/>
              <faultSequence/>
          </resource>
          <resource methods="POST" uri-template="/readRecord">
              <inSequence>
                <property expression="json-eval($.sysId)" name="sysId" scope="default" type="STRING"/>
                  <sequence key="ReadRecord"/>
                  <respond/>
              </inSequence>
              <outSequence/>
              <faultSequence/>
          </resource>
      </api>

    ```

{!references/connectors/exporting-artifacts.md!}

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}

## Testing

### Post Record Operation

> 1. Create a file called data.json with the following payload. You can further refer to the parameters from [here.](https://docs.servicenow.com/bundle/orlando-application-development/page/integrate/inbound-rest/concept/c_TableAPI.html#c_TableAPI) 

Reviewer 2: ```suggestion
1. Create a file called data.json with the following payload. You can further refer to the parameters from [here](https://docs.servicenow.com/bundle/orlando-application-development/page/integrate/inbound-rest/concept/c_TableAPI.html#c_TableAPI).
```

Author: Done
